utils.py

```diff
@@ -456,6 +471,452 @@ def calc_hotteling_statistic_for_sample(observed_features: np.ndarray, sample_fe
+    Parameters
+    ----------
+    metrics_collection
+        The `MetricsCollection` with relation to which the optimization is carried out.
+        # TODO: we can't add a type hint for this, due to circular import (utils can't import from metrics, as metrics
+            already imports from utils). This might suggest that this isn't the right place for this function.
+    observed_network
+        The observed network used as data for the optimization.
+    initial_thetas
+        The initial vector of parameters. If `None`, the initial state is randomly sampled from (0,1)
+    lr
+        The learning rate of the gradient-ascent, scales the step size of the optimization.
+    max_iter
+        The maximum number of optimization iterations to run.
+    stopping_thr
+        The fraction of change in the objective function (which is the log-likelihood) that is used as stopping
+        criterion (i.e., if the percent change of the objective is smaller than this threshold we stop)
+    is_distributed
+        Whether the calculations are carried locally or distributed over many compute nodes of an IBM LSF cluster.
+    Returns
+        Parameters of the trained model.
+    -------
+
+    """
+    num_features = metrics_collection.calc_num_of_features()
+    if initial_thetas is None:
+        thetas = np.random.rand(num_features, 1)
+    else:
+        thetas = initial_thetas.copy()
+    cur_log_like = -np.inf
+    prev_log_like = -np.inf
+    if not is_distributed:
+        Xs, ys = metrics_collection.prepare_mple_data(observed_network)
+    else:
+        out_dir_path = (Path.cwd().parent / "OptimizationIntermediateCalculations").resolve()
+        data_path = (out_dir_path / "data").resolve()
+        os.makedirs(data_path, exist_ok=True)
+
+        # Copy the `MetricsCollection` and the observed network to provide its path to children jobs, so they will be
+        # able to access it.
+        metric_collection_path = os.path.join(data_path, 'metric_collection.pkl')
+        with open(metric_collection_path, 'wb') as f:
+            pickle.dump(metrics_collection, f)
+        observed_net_path = os.path.join(data_path, 'observed_network.pkl')
+        with open(observed_net_path, 'wb') as f:
+            pickle.dump(observed_network, f)
+
+    idx = 0
+    with objmode(start='f8'):
+        start = time.perf_counter()
+    print("Logistic regression optimization started")
+    sys.stdout.flush()
+    for i in range(max_iter):
+        idx = i
+        if not is_distributed:
+            prediction, cur_log_like, grad, hessian = local_mple_logistic_regression_optimization_step(Xs, ys, thetas)
+        else:
+            prediction, cur_log_like, grad, hessian = distributed_logistic_regression_optimization_step(data_path,
+                                                                                                        thetas)
+        if (i - 1) % 1 == 0:
```

Comment:
won't this always evaluate to True?

Reply:
No, it is passed here as a field of ERGM:
https://github.com/tomtalp/pyERGM/blob/4221e677f7cb97ca81d27bdae0e2ac259e8c403c/ergm.py#L172
It can be initialized for distributed optimization.